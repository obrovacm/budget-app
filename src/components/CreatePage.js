import React from "react";

const CreatePage = () => <h3>Create Page</h3>;

export default CreatePage;
