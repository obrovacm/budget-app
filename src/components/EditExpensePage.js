import React from "react";

const EditExpensePage = props => {
  console.log(props);

  return (
    <div>
      <h3>Edit Page with id: {props.match.params.id}</h3>
      <p>
        Router gives additional info in props object, such as this ID, from URL.
      </p>
    </div>
  );
};

export default EditExpensePage;
