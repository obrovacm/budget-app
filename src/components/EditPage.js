import React from "react";

const EditPage = () => <h3>Edit Page</h3>;

export default EditPage;
