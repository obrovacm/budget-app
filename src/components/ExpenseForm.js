import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      submitMessage: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();
    let message;

    if (!this.state.description || !this.state.amount) {
      message = "Error: Please provide description and amount.";
    } else {
      message = "Expense added.";
      this.props.onSubmit({
        description: this.state.description,
        // converting string to float
        amount: parseFloat(this.state.amount, 10),
        // taking value in milliseconds (so it's possible to compare it)
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }

    this.setState(() => ({
      submitMessage: message
    }));
  };

  render() {
    return (
      <div>
        {!!this.state.submitMessage && <p>{this.state.submitMessage}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            firstDayOfWeek={1}
            isOutsideRange={() => false}
            hideKeyboardShortcutsPanel={true}
            displayFormat={"DD.MM.YYYY."}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button type="submit">Save Expense</button>
        </form>
      </div>
    );
  }
}
