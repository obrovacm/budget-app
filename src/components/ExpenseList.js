import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses"; // Get visible/filtered expenses

const ExpenseList = props => (
  <div className="expense-list">
    {props.expenses.map(expense => (
      <ExpenseListItem {...expense} key={expense.id} />
    ))}
  </div>
);
// {...expense} spreads an object as props

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

//higher order component
export default connect(mapStateToProps)(ExpenseList);
