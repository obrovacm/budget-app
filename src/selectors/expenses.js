import moment from "moment";

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdMoment = moment(expense.createdAt);

      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdMoment, "day")
        : true;

      const textMatchNote =
        typeof text !== "string" ||
        expense.note.toLowerCase().includes(text.toLowerCase());
      const textMatchDescription =
        typeof text !== "string" ||
        expense.description.toLowerCase().includes(text.toLowerCase());

      return (
        startDateMatch &&
        endDateMatch &&
        (textMatchNote || textMatchDescription)
      );
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      // else if (sortBy === "amount") {
      //   return a.amount < b.amount ? 1 : -1;
      // }

      // this way there's no 'default' return warning
      return a.amount < b.amount ? 1 : -1;
    });
};
