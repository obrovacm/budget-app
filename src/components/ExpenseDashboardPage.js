import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div>
    <hr />
    <h3>Filter Expense List</h3>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default ExpenseDashboardPage;
