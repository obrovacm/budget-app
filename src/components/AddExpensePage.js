import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

const AddExpensePage = props => (
  <div>
    <hr />
    <h3>Add Expense</h3>
    <ExpenseForm
      onSubmit={expense => {
        // dispatching expense outside of the form provides possibility
        // to reuse component for another function, such as editing expenses
        props.dispatch(addExpense(expense));
        // redirects user to home page and also resets the form
        // since it's component will dismount
        props.history.push("/");
      }}
    />
  </div>
);

export default connect()(AddExpensePage);
