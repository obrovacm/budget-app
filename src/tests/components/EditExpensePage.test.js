import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;
let expense1 = expenses[0];

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      expense={expense1}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  );
});

test("should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expense1);
  expect(history.push).toHaveBeenCalledWith("/");
  expect(editExpense).toHaveBeenCalledWith(expense1.id, expense1);
});

test("should handle removeExpense", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenCalledWith("/");
  expect(removeExpense).toHaveBeenCalledWith({ id: expense1.id });
});
