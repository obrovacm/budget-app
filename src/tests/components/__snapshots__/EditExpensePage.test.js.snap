// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <hr />
  <h3>
    Edit Expense
  </h3>
  <ExpenseForm
    expense={
      Object {
        "amount": 9999,
        "createdAt": -345600000,
        "description": "first",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
